library/pkcs7_write.c

```diff
@@ -0,0 +1,829 @@
+    rc = pkcs7_write_data( start, size, ptr, WRITE_SIGNATURE, signature,
+                      sig_size, 0 );
+    if( rc ) 
+    {
+        mbedtls_printf(
+            "Failed to add signature to PKCS7 (signature generation was successful "
+            "however)\n" );
+    }
+out:
+    mbedtls_pk_free( priv_key );
+    if( priv_key ) mbedtls_free( priv_key );
+    if( hash ) mbedtls_free( hash );
+    if( signature ) mbedtls_free( signature );
+    return ( rc );
+}
+
+static int pkcs7_set_algorithm_ids( unsigned char **start, size_t *size,
+                           unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info,
+                           mbedtls_x509_crt *pub, unsigned char *priv,
+                           size_t priv_size )
+{
+    int rc;
+    char *sig_type = NULL;
+
+    rc = pkcs7_set_signature( start, size, ptr, pkcs7_info, pub, priv, priv_size );
+    if( !rc ) 
+    {
+        // make sure it is rsa encryption, that is all we support right now
+        sig_type = (char *) pub->pk.pk_info->name;
+        if( strcmp( sig_type, "RSA" ) ) 
+        {
+            rc = MBEDTLS_ERR_PKCS7_INVALID_CERT;
+            mbedtls_printf( "ERROR: Public Key is of type %s expected RSA\n", sig_type );
+            return rc;
+        }
+        rc = pkcs7_write_data(
+                 start, size, ptr, WRITE_OID, (void *) MBEDTLS_OID_PKCS1_RSA, 
+                 strlen( MBEDTLS_OID_PKCS1_RSA ), 0 );
+        if( !rc ) 
+        {
+            rc = pkcs7_write_data(
+                  start, size, ptr, WRITE_OID, (void *) pkcs7_info->hash_funct_oid, 
+                  strlen( pkcs7_info->hash_funct_oid ), 0 );
+        }
+    }
+    return ( rc );
+}
+
+static int pkcs7_set_signer_cert_data( unsigned char **start, size_t *size,
+                             unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info,
+                             mbedtls_x509_crt *pub, unsigned char *priv,
+                             size_t priv_size )
+{
+    int rc, signed_info_version = 1;
+    size_t bytes_written_in_step, curr_len;
+    unsigned char tag = MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE;
+    rc = pkcs7_set_algorithm_ids( start, size, ptr, pkcs7_info, pub, priv, priv_size );
+    if( !rc ) 
+    {
+        // add serial
+        curr_len = *size - ( *ptr - *start );
+        rc = pkcs7_write_data( start, size, ptr,
+                          WRITE_SERIAL,
+                          pub->serial.p, pub->serial.len, 0 );
+        if( !rc ) 
+        {
+            rc = pkcs7_write_data( start, size, ptr, WRITE_RAW_BUFFER,
+                              pub->issuer_raw.p, pub->issuer_raw.len, 0 );
+            if( !rc ) 
+            {
+                // add info
+                bytes_written_in_step = *size - ( *ptr - *start ) - curr_len;
+                rc = pkcs7_write_data( start, size, ptr,
+                                  WRITE_TAG, &tag,
+                                  bytes_written_in_step, 0 );
+                if( !rc ) 
```

Comment:
Style: is nesting the best choice here? Is there a reason not to follow the same "exit early on error" pattern as other functions?